### PR TITLE
Document setting up a gps constellation

### PR DESCRIPTION
**Pull request by Lupus590**
related: #1068

### CODE REVIEW

File: doc/guides/gps_setup.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+Before GPS can be used in a dimention (e.g. the Nether) a GPS host constellation needs to be set up. Traditionally this is done near the max build height for maximum range, but ender modems can beat normal modes in range even from bedrock.
+
+In order to give the best results, a GPS constallation needs atleast four computers. One can theroetically get away with three for most GPS requests, but four changes that to 100% accuracy when the requester is within all of their ranges, provided that they are placed correctly. More than four GPS hosts per constallation is redundent, but it does not harm.
+
+## Building a GPS constellation
+![An example GPS constellation.](images/gps-constellation-example.png){.big-image}
+
+Assuming that you are using four computers per costallation, you will want each computer that makes up the constallation to be out of plane with atleast one other computer in the constalation. To put this another way, if you can make a straight wall or floor (diagonal also counts as straight, having a corner or curve doesn't) that touches all four computers then you need to move one of the computers so that it's nolonger in contact with the wall/floor.
```
Comment (DVD-DAVIDE) [inline doc/guides/gps_setup.md:19]: Typos on row  19:
"constallation" instead of "constellation" (twice)
"constalation" instead of "constellation"

File: doc/guides/gps_setup.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+Having the GPS constallation computers within a few blocks of each other is what people commonly do but is not required. So long as atleast four GPS hosts can respond to the requester the request will be fulfilled.
```
Comment (DVD-DAVIDE) [inline doc/guides/gps_setup.md:21]: Typo on row 21:
"constallation" instead of "constellation"

File: doc/guides/gps_setup.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+GPS only works with wireless modems, ender modems count as wireless. It doesn't matter what side you attach them to, just that each computer in the constellation has one.
+
+:::note Ender modems
+You might be aware that ender modems have a very large range, this makes them very usesful for setting up GPS hosts. Infact, if you do this then you will likely only need one GPS constellation for the whole dimention (in vanilla this will mean that you'd need three constellation, one for the Overworld, one for the Nether, and one for the End).
```
Comment (DVD-DAVIDE) [inline doc/guides/gps_setup.md:26]: Typo on row 26:
"usesful" instead of useful
"dimention" instead of "dimension"
"constellation" instead of "constellations"

File: doc/guides/gps_setup.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# Setting up GPS
+Some quick definitions for this page.
+```
+GPS host: a computer running the GPS program in host mode
+GPS constellation: a group of GPS hosts
+```
+
+Before GPS can be used in a dimention (e.g. the Nether) a GPS host constellation needs to be set up. Traditionally this is done near the max build height for maximum range, but ender modems can beat normal modes in range even from bedrock.
+
+In order to give the best results, a GPS constallation needs atleast four computers. One can theroetically get away with three for most GPS requests, but four changes that to 100% accuracy when the requester is within all of their ranges, provided that they are placed correctly. More than four GPS hosts per constallation is redundent, but it does not harm.
```
Comment (DVD-DAVIDE) [inline doc/guides/gps_setup.md:14]: Typos on line 14:
"atleast" instead of "at least"
"theroetically" instead of "theoretically"
"constallation" instead of "constellation"
"redundent" instead of "redundant"

File: doc/guides/gps_setup.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+:::note Why use MC's coordinates?
+CC doesn't actually care if you use Minecraft's coordinate system, so long as all of the GPS host with overlapping ranges use the same reference point (requesting computers will get confused if hosts have different reference points). However, using MC's coordinate system does provide a nice standard to adopt server wide. It also is good to use as command computers use it when they cheat to find their position with @{gps.locate}.
+:::
+
+:::note Modem messages come from the computer's position not the modem's
+Wireless modems transmit from the block that they are attached to *not* the block space that they occupy, the coordinates that you input into your GPS host should be the postion of the computer and not the position of the modem.
+:::
+
+## A brief look at how the GPS API works in CC
+With the exception of command computers and user code overrides, ComputerCraft actually mimics real life GPS; computers triagulate themselves using signals with known sender positions (which is why we have to tell the GPS hosts where they are). This means that the average CC computer will need a wireless modem and be in range of a GPS constellation.
```
Comment (DVD-DAVIDE) [inline doc/guides/gps_setup.md:51]: Error on line 51:
Computers do not use triangulation, they use trilateration.

File: doc/guides/gps_setup.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+:::note Double duty
+You could also set up one of their GPS hosts (per constellation) to run the rednet `repeat` program via @{multishell} or the @{parallel} API. Do note that if you use the parallel API that the console output of the two programs will merge together, this is harmless but can make it harder to understand the output.
```
Comment (DVD-DAVIDE) [inline doc/guides/gps_setup.md:54]: To add on line 54:
Explain the purpose of the repeat program.

**Review by DVD-DAVIDE — CHANGES_REQUESTED** [review]

File: doc/guides/gps_setup.md (hunk 6 above)
Comment (Lupus590) [reply doc/guides/gps_setup.md:54]: I feel like explaining what the repeat program does is out of scope. Going to leave this open to see what others thing though.

Comment (Lupus590) [reply doc/guides/gps_setup.md]: @SquidDev What are your thoughts on this section?

File: doc/guides/gps_setup.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+GPS host: a computer running the GPS program in host mode
+GPS constellation: a group of GPS hosts
+```
+
+Before GPS can be used in a dimension (e.g. the Nether) a GPS host constellation needs to be set up. Traditionally this is done near the max build height for maximum range, but ender modems easily beats a normal modem's max range at any height.
+
+To give the best results, a GPS constellation needs at least four computers. One can theoretically get away with three for most GPS requests, but four changes that to 100% accuracy when the requester is within all of their ranges, provided that they are placed correctly. Having more than four GPS hosts per constellation is redundant, but it does not harm.
+
+## Building a GPS constellation
+![An example GPS constellation.](images/gps-constellation-example.png){.big-image}
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]:
```suggestion
![An example GPS constellation.](/images/gps-constellation-example.png){.big-image}
```

As this is located within a sub-directory of the site.

File: doc/guides/gps_setup.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
+---
+module: [kind=guide] gps_setup
+---
+
+# Setting up GPS
+Some quick definitions for this page.
+```
+GPS host: a computer running the GPS program in host mode
+GPS constellation: a group of GPS hosts
+```
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: I think this is probably better done as prose. So something like:

> The @{gps} API allows computers and turtles to find their current position using wireless modems.
> 
> In order to use GPS, you'll need to set up multiple _GPS hosts_. These are computers running the special `gps host` program, which tell other computers the host's position. Several hosts running together are known as a _GPS constellation._

Obviously feel free to tweak/rewrite this, just trying to give an idea :).

File: doc/guides/gps_setup.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+Before GPS can be used in a dimension (e.g. the Nether) a GPS host constellation needs to be set up. Traditionally this is done near the max build height for maximum range, but ender modems easily beats a normal modem's max range at any height.
+
+To give the best results, a GPS constellation needs at least four computers. One can theoretically get away with three for most GPS requests, but four changes that to 100% accuracy when the requester is within all of their ranges, provided that they are placed correctly. Having more than four GPS hosts per constellation is redundant, but it does not harm.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: > One can theoretically get away with three for most GPS requests,

Hrm, are you sure about this? My understanding is that you need four - three only narrows your position down to two points.

File: doc/guides/gps_setup.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+CC doesn't care if you use Minecraft's coordinate system, so long as all of the GPS hosts with overlapping ranges use the same reference point (requesting computers will get confused if hosts have different reference points). However, using MC's coordinate system does provide a nice standard to adopt server-wide. It also is consistaint with how command computers get their location, they use MC's command system to get their block which returns that in MC's coordinate system.
+:::
+
+:::note Modem messages come from the computer's position, not the modem's
+Wireless modems transmit from the block that they are attached to *not* the block space that they occupy, the coordinates that you input into your GPS host should be the position of the computer and not the position of the modem.
+:::
+
+## A brief look at how the GPS API works in CC
+Except for command computers and user code overrides, ComputerCraft mimics real-life GPS; computers trilaterate themselves using signals with known sender positions (which is why we have to tell the GPS hosts where they are). This means that the average CC computer will need a wireless modem and be in range of a GPS constellation.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: What do you mean by "the average CC computer"? I assume you're referring to non-command ones here?

**Review by SquidDev — CHANGES_REQUESTED** [review]

File: doc/guides/gps_setup.md (hunk 6 above)
Comment (SquidDev) [reply doc/guides/gps_setup.md]: Yeah, I think this paragraph could probably be dropped entirely.

File: doc/guides/gps_setup.md (hunk 10 above)
Comment (Lupus590) [reply doc/guides/gps_setup.md]: Yeah, non-command computers is what I'm trying to say.

File: doc/guides/gps_setup.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+GPS only works with wireless modems, ender modems count as wireless. It doesn't matter what side you attach them to, just that each computer in the constellation has one.
+
+:::note Ender modems
+You might be aware that ender modems have a very large range, this makes them very useful for setting up GPS hosts. If you do this then you will likely only need one GPS constellation for the whole dimension (in vanilla this will mean that you'd need three constellations, one for the Overworld, one for the Nether, and one for the End).
+
+Also, don't worry about an ender modem's ability to work across dimensions, the GPS API is smart enough to ignore constellations that are not in its dimension.
+:::
+
+## Configureing the constellation
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]:
```suggestion
## Configuring the constellation
```

File: doc/guides/gps_setup.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+It doesn't matter what kind of startup file you use, so long as it's a valid one; the contents of the file is going to be the same.
+
+Here's a template for the startup file's contents:
+```lua
+shell.run("gps", "host", x, y, z)
+```
+Where `x`, `y`, and `z` are the respective coordinates of the CC computer. The easiest way to get the computer's coordinates is to look at it and press `F3`, its coordinates can then be found on the right side of the screen as `Targeted Block`.
+
+:::note Why use MC's coordinates?
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: Replace MC with Minecraft.

I think it's also worth moving this into the "A brief look at how the GPS API works in CC". I think the actual guide bit needs to be as simple as possible.

File: doc/guides/gps_setup.md (hunk 10 above)
Comment (SquidDev) [reply doc/guides/gps_setup.md]: Yeah, I'd just leave that off then.

I'm not really sure about this section in its current state. Right now it really boils down to "it's Trilateration!". I don't know if we want a couple of sentences explanation, or just drop it entirely.

File: doc/guides/gps_setup.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
+---
+module: [kind=guide] gps_setup
+---
+
+# Setting up GPS
+The @{gps} API allows computers and turtles to find their current position using wireless modems.
+
+In order to use GPS, you'll need to set up multiple GPS hosts. These are computers running the special gps host program, which tell other computers the host's position. Several hosts running together are known as a GPS constellation.
+
+Before GPS can be used in a dimension (e.g. the Nether) a GPS host constellation needs to be set up. Traditionally this is done near the max build height for maximum range, but ender modems easily beats a normal modem's max range at any height.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: The first bit of this kinda duplicates the previous paragraph now.

I think I'd probably drop this paragraph (and the one below) and then mention these considerations in the "Building a GPS constellation" section.

File: doc/guides/gps_setup.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+:::note Ender modems
+You might be aware that ender modems have a very large range, this makes them very useful for setting up GPS hosts. If you do this then you will likely only need one GPS constellation for the whole dimension (in vanilla this will mean that you'd need three constellations, one for the Overworld, one for the Nether, and one for the End).
+
+Also, don't worry about an ender modem's ability to work across dimensions, the GPS API is smart enough to ignore constellations that are not in its dimension.
+:::
+
+## Configureing the constellation
+To provide computers (or more commonly, turtles) their location, every GPS host needs to know its own location. GPS host can get this from other constellations, but we are going to input it manually, which means that we are going to make a startup file.
+
+It doesn't matter what kind of startup file you use, so long as it's a valid one; the contents of the file is going to be the same.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]:
```suggestion
```

I think we want to make this as beginner friendly as possible, so even mentioning the kinds of startup files is out of scope.

**Review by SquidDev — CHANGES_REQUESTED** [review]
Sorry, been thinking about this a chunk more, and I think I'm going to propose another awkward restructuring. Content is good, it's just a little hard to follow along right now!

I think the main thing we want to do is merge the "Building a GPS constellation" and "Configuring the constellation" sections and then make them as step-by-step as possible. Maybe even write this as a list (not sure, that might be a bit too much!)

So literally something like:
 1. You will need: (4 computers, 4 wireless modems). Maybe worth briefly mentioning the ender/wireless tradeoff here. Maybe mention you'll need wireless modems (but either kind is fine) for each computer which needs to find its position with GPS.

 2. Find a place to build (talk about the size requirements, _maybe_ mention it's better to be build with non-ender modems higher up (though I'm not sure this is true on 1.18, haven't done the maths with the new world height!)

 3. Place down one computer and modem, talk about F3, write the startup file with them.

 4. Talk about the next three computers. Honestly, I'd be much more prescriptive about this (get them to build the _exact_ layout you've got in the screenshot).

Then maybe get them to test this with another computer and confirm the position is correct.

---

I don't know! I'm obviously not a beginner here, and haven't ever used the built-in GPS server (always rolled my own >:D), so do feel free to say if you disagree!

**Comment by Lupus590** [comment]
I decided to keep the two sections just in case someone is reading the guide because they remember one half but not the other.

Building it is quite different to configuring it so I figured that I'd keep the distinction.

File: doc/guides/gps_setup.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+---
+
+# Setting up GPS
+The @{gps} API allows computers and turtles to find their current position using wireless modems.
+
+In order to use GPS, you'll need to set up multiple GPS hosts. These are computers running the special gps host program, which tell other computers the host's position. Several hosts running together are known as a GPS constellation.
+
+In order to give the best results, a GPS constellation needs at least four computers. More than four GPS hosts per constellation is redundant, but it does not harm.
+
+A computer needs a wireless modem and to be in range of a GPS constellation to use the GPS API. The reason for this is that ComputerCraft mimics real-life GPS by making use of the distance parameter of @{modem_message|modem messages} and some maths.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: I think move this into the wireless vs ender modem section.

File: doc/guides/gps_setup.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+Additionally, the GPS constellation needs to be chunk loaded. CC doesn't provide any chunk loading capabilities so you'll need to make use of another mod or the vanilla `forceload` command.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: Similarly, move this into the "picking a location" section. It might also be worth mentioning that it's best to build it all in one chunk in order to simplify loading.

File: doc/guides/gps_setup.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+## Building a GPS constellation
+![An example GPS constellation.](/images/gps-constellation-example.png){.big-image}
+
+We are going to build our GPS constellation as shown in the image above. You will need 4 computers and either 4 wireless modems or 4 ender modems - try not to mix ender and wireless modems together as you might get some odd behavior when your requesting computers are out of range.
+
+:::note Ender modems vs wireless modems
+You might be aware that ender modems have a very large range, this makes them very useful for setting up GPS hosts. If you do this then you will likely only need one GPS constellation for the whole dimension (in vanilla this will mean that you'd need three constellations, one for the Overworld, one for the Nether, and one for the End).
+
+Also, don't worry about an ender modem's ability to work across dimensions, the GPS API is smart enough to ignore constellations that are not in its dimension.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: I think we can remove this paragraph.

File: doc/guides/gps_setup.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+As for the actual building itself, place your first computer in one of the corners of your 6 by 6 by 6 space. Remember which computer this is as you other computers need to be placed relative to it. For your second computer, place it 4 blocks above the first computer leaving 3 blocks of air between them and one block of air above the second computer that is within your 6 by 6 by 6 space. Go back to your first computer and place your third computer 5 blocks in front of your first computer, leaving 4 blocks of air between them. Finally for the fourth computer, go back to your first computer and place it 5 blocks right of your first computer, leaving 4 blocks of air between them.
+
+With all four computers placed within the 6 by 6 by 6 space, place one modem on top of each computer. Yuo should have 4 modems and 4 computers all within your 6 by 6 by 6 space where each modem is attached to a computer and each computer has a modem.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]:
```suggestion
With all four computers placed within the 6 by 6 by 6 space, place one modem on top of each computer. You should have 4 modems and 4 computers all within your 6 by 6 by 6 space where each modem is attached to a computer and each computer has a modem.
```

File: doc/guides/gps_setup.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+If you do use wireless modems then you may find that you need multiple GPS constellations to cover your needs.
+:::
+
+Locate where you want to place your GPS constellation, you will need at least a 6 by 6 by 6 blocks of space. If you are using wireless modems then you may want to build your constellation as high as you can because high altitude boosts modem message range and thus the radius that your constellation covers.
+
+As for the actual building itself, place your first computer in one of the corners of your 6 by 6 by 6 space. Remember which computer this is as you other computers need to be placed relative to it. For your second computer, place it 4 blocks above the first computer leaving 3 blocks of air between them and one block of air above the second computer that is within your 6 by 6 by 6 space. Go back to your first computer and place your third computer 5 blocks in front of your first computer, leaving 4 blocks of air between them. Finally for the fourth computer, go back to your first computer and place it 5 blocks right of your first computer, leaving 4 blocks of air between them.
```
Comment (SquidDev) [inline doc/guides/gps_setup.md]: "6 by 6 by 6" is a little clunky, but I don't know if "6x6x6" is any clearer. Both are a little blughr

**Review by SquidDev — CHANGES_REQUESTED** [review]
Thank you! Just a couple of small things.

File: doc/guides/gps_setup.md (hunk 19 above)
Comment (Lupus590) [reply doc/guides/gps_setup.md]: Maybe say it longer the first time with 6x6x6 in brackets and then use the latter the rest of the time?

> ...you will a space 6 blocks in width, 6 blocks in length, and 6 blocks high (6x6x6).
> ...place your first computer in one of the corners of your 6x6x6.
> ...one block of air above the second computer that is within your 6x6x6.

**Comment by SquidDev** [comment]
Really sorry for taking so long to look at this again, thank you so much for your patience in my many requests! Merged manually in d6afee8deb3dbd215a9f8d08c25e4f375d4ff82d.
